fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java

```diff
@@ -135,56 +168,82 @@ public static ScalarOperator translate(Expr expression, ExpressionMapping expres
 
     public static ScalarOperator translate(Expr expression) {
         IgnoreSlotVisitor visitor = new IgnoreSlotVisitor();
-        return visitor.visit(expression, null);
+        return visitor.visit(expression, new Context());
     }
 
-    public static ScalarOperator translateWithoutRewrite(Expr expression, ExpressionMapping expressionMapping,
-                                                         ColumnRefFactory columnRefFactory) {
-        ColumnRefOperator columnRefOperator = expressionMapping.get(expression);
-        if (columnRefOperator != null) {
-            return columnRefOperator;
-        }
+    private static final class Context {
 
-        Visitor visitor = new Visitor(expressionMapping, new ArrayList<>(), columnRefFactory);
-        ScalarOperator result = visitor.visit(expression, null);
+        public final boolean hasSubquery;
+        public final boolean useSemiAnti;
+        public final List<Expr> outerExprs;
 
-        ScalarOperatorRewriter scalarRewriter = new ScalarOperatorRewriter();
-        result = scalarRewriter.rewrite(result, ScalarOperatorRewriter.DEFAULT_TYPE_CAST_RULE);
+        public Context() {
+            this.hasSubquery = false;
+            this.useSemiAnti = false;
+            this.outerExprs = Collections.emptyList();
+        }
 
-        requireNonNull(result, "translated expression is null");
-        return result;
+        public Context(boolean hasSubquery, boolean useSemiAnti, List<Expr> outerExprs) {
+            this.hasSubquery = hasSubquery;
+            this.useSemiAnti = useSemiAnti;
+            this.outerExprs = outerExprs;
+        }
+
+        public Context clone(Expr node) {
+            if (!hasSubquery) {
+                return this;
+            }
+            List<Expr> outerExprs = Collections.emptyList();
+            if (node.getChildren().stream().filter(Subquery.class::isInstance).count() == 1) {
```

Comment:
can you add some comment to explain why only consider node has only one subquery?

Reply:
This conditionn is weird, and I changed it.